### ISSUE TITLE
Regressions from #5427 (FBO_Fixes)

### ISSUE THREAD

**Issue by jenci8888** [RPCS3/rpcs3#5514]
**Issue#1:** I'm getting this error which I am unable go past intro logo screen while OGL makes go ingame.
**Issue#2:** Graphical issues NFS Rivals, therefore WCB enabled fixes it. Thus harder to run which game freezes loading screen before "start button" screen, I tried to capture the "RSX Capture" but I am unable to do it. (See log)

**Game:** Sonic All-Star Racing Transformed
`F {RSX [0x160cdb0]} RSX: class std::runtime_error thrown: Assertion Failed! Vulkan API call failed with unrecoverable error: Device lost (Driver crashed with unspecified error or stopped responding and recovered) (VK_ERROR_DEVICE_LOST) (in file c:\projects\rpcs3\rpcs3\emu\rsx\vk\vkhelpers.h:1275)`

[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/2730369/RPCS3.zip)

**Game:** Need For Speed: Rivals
**Before:** WCB disabled (Doesn't matter any build)
![image](https://user-images.githubusercontent.com/40495799/50737880-f6e9dc00-11cd-11e9-90f6-830795e824e8.png)

**After:** WCB disabled
![image](https://user-images.githubusercontent.com/40495799/50737858-bb4f1200-11cd-11e9-9d85-31cdcfdb15ad.png)

**After#2:** WCB enabled
![image](https://user-images.githubusercontent.com/40495799/50737875-ec2f4700-11cd-11e9-853e-d121d84cc504.png)

[RPCS3_NFS_Rivals.log.gz](https://github.com/RPCS3/rpcs3/files/2730584/RPCS3_NFS_Rivals.log.gz)

I provided log. Feel free post here to make regressions.

**Comment by stride21** [comment]
Prior to #5427 with Async Shader Compiler enabled:
![unc previous](https://user-images.githubusercontent.com/37915864/50848465-f53d2700-1339-11e9-8369-999393aa72d1.JPG)
#5427 with Async Shader Compiler enabled:
![unc pr](https://user-images.githubusercontent.com/37915864/50848478-ff5f2580-1339-11e9-9b13-7afc9a86cbb4.JPG)
#5427 with Async Shader Compiler disabled:
![disabled](https://user-images.githubusercontent.com/37915864/50848557-2fa6c400-133a-11e9-916a-813671fa24ff.JPG)
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2737909/RPCS3.log.gz)

Prior to #5427 with Async Shader Compiler enabled:
![previous last of us](https://user-images.githubusercontent.com/37915864/50849211-eeafaf00-133b-11e9-9678-1372cf66d7a1.JPG)
#5427 with Async Shader Compiler enabled:
![last of us pr](https://user-images.githubusercontent.com/37915864/50849110-a85a5000-133b-11e9-9646-8a4fadf442dc.JPG)
#5427 with Async Shader Compiler disabled:
![last of us master](https://user-images.githubusercontent.com/37915864/50849188-dc357580-133b-11e9-9633-cc833658ae65.JPG)
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2737910/RPCS3.log.gz)

**Comment by kd-11** [comment]
Likely because I decided to force data collapse even if a write did not happen, which somehow breaks async compiling because that is itself a nasty hack. I'll just return it to how it was before, although I wonder if that will also bring in regressions of its own.

**Comment by cubatilles** [comment]
Persona 4 arena and persona 4 arena ultimax seem to have a regression too due to this PR. The pink flickering glitch on the characters is back, no matter if you enable WCB or strict rendering mode.

Sorry I'm not on my main computer now and cannot post screenshots.

**Comment by Asinin3** [comment]
Make sure you have anisotropic filtering set to Automatic for Persona 4 Arena. I'll check it in a sec.

**Comment by cubatilles** [comment]
Of course its on automatic. Same config on prefbo fixes builds just wotks ok

**Comment by stride21** [comment]
Seems to be affecting Battlefield: Bad Company as well.
Async Shader Compiler enabled:
![capture](https://user-images.githubusercontent.com/37915864/50980083-81c42280-14bd-11e9-926d-5a89e716069e.JPG)
Async Shader Compiler disabled:
![ffff](https://user-images.githubusercontent.com/37915864/50980091-838de600-14bd-11e9-8b91-01e9fb890d36.JPG)


[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2746068/RPCS3.log.gz)

**Comment by kd-11** [comment]
The async 'regression' is literally a one-liner hack. Don'r worry about it too much, it will be submitted with the next PR, likely in the next 2 days.

**Comment by Xcedf** [comment]
Seeing now that only async shaders are broken, tried out Async Shaders Off with TLoU
and shadowmaps work beautyfully
![1](https://user-images.githubusercontent.com/7060586/51062490-1a66b980-1608-11e9-8c1f-304c57618305.jpg)

**Comment by Asinin3** [comment]
Assassins Creed Rogue now crashes with OpenGL after pressing start on the press start screen 5seconds after it boots. Very easy to reproduce.
`F {RSX [0x042f764]} RSX: class std::runtime_error thrown: Unhandled GL format 0x8058 (in file c:\projects\rpcs3\rpcs3\emu\rsx\gl\glhelpers.h:1611)`
[RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2755455/RPCS3.log.gz)

**Comment by Emulator-Team-2** [comment]
[ Grand Theft Auto IV ]
GPU: AMD RX 570 / 19.1.1
RSX-Capture: https://mega.nz/#!kVR1hKyI!GnGoVjlD2r1tYNOOxmc5DfKyd0LiJu9glABummzrkts
Log: [RPCS3.log.gz](https://github.com/RPCS3/rpcs3/files/2770029/RPCS3.log.gz)

v0.0.5-7731 / v0.0.5-7657 
![1](https://user-images.githubusercontent.com/38318848/51339146-fa157f80-1a8b-11e9-846f-6287f3a0d1d4.png)

**Comment by kd-11** [comment]
@Emulator-Team-2 Retest with https://github.com/RPCS3/rpcs3/pull/5565 for GTA

**Comment by Emulator-Team-2** [comment]
[6453d70d]
No change, this game still looks the same.

**Comment by kd-11** [comment]
Is it vulkan-only? You did not mention if OGL was also affected.

**Comment by Emulator-Team-2** [comment]
v0.0.5-7742 / [OpenGL]  
![1](https://user-images.githubusercontent.com/38318848/51605199-a339fb00-1f0e-11e9-85b4-b8129de2635a.png)

--------------------------------------------------------

[rsx: Fixups and improvements]
On your new pr, GTA 4 has a black screen and crashes a few seconds later after the loading screen with this error.
```
F {RSX [0x029dc44]} RSX: class std::runtime_error thrown: Verification failed: 
(in file c:\projects\rpcs3\rpcs3\emu\rsx\gl\glvertexbuffers.cpp:245)]
```
[RPCS3.log](https://github.com/RPCS3/rpcs3/files/2787108/RPCS3.log)

**Comment by kd-11** [comment]
Does vulkan also crash on the PR build? Its important to check both when reporting bugs/regressions
.

**Comment by Emulator-Team-2** [comment]
No, the game doesn't crash with Vulkan, but the graphics are completely broken.

**Comment by creeperjedi** [comment]
Tom Clancy's H.A.W.X [BLUS30186] no longer displays anything, only a black screen.

Settings: Default CPU Decoder settings and both renderers work. Choice of renderer doesn't make a difference, but different renderers were used for the RSX Captures below because in each case, the other renderer would crash the game and be unable to produce a capture.

Also, ignore the FPS counter on both screenshots as they are not indicative of average performance.
<hr>
Before #5427:
<img width="641" alt="h a w x screenshot - 0 0 5-7658" src="https://user-images.githubusercontent.com/38361979/51638991-d6d84d80-1f2d-11e9-94db-ba2de1c87e81.png">

[H.A.W.X. RSX Capture (OpenGL) - 0.0.5-7658.zip](https://github.com/RPCS3/rpcs3/files/2789304/H.A.W.X.RSX.Capture.OpenGL.-.0.0.5-7658.zip)
[H.A.W.X. Log - 0.0.5-7658.log](https://github.com/RPCS3/rpcs3/files/2789307/H.A.W.X.Log.-.0.0.5-7658.log)
<hr>
From #5427 to current master:
<img width="641" alt="h a w x screenshot - 0 0 5-7678" src="https://user-images.githubusercontent.com/38361979/51640142-b52c9580-1f30-11e9-9af0-bbf7301d2e68.png">

[H.A.W.X. RSX Capture (Vulkan) - 0.0.5-7678.zip](https://github.com/RPCS3/rpcs3/files/2789342/H.A.W.X.RSX.Capture.Vulkan.-.0.0.5-7678.zip)
[H.A.W.X. Log - 0.0.5-7678.log](https://github.com/RPCS3/rpcs3/files/2789344/H.A.W.X.Log.-.0.0.5-7678.log)

**Comment by kd-11** [comment]
All these regressions need testing on the current PR https://github.com/RPCS3/rpcs3/pull/5565.
HAWX for example displays correct graphics in the PR build.

**Comment by kd-11** [comment]
I think all these issues have been resolved after the PR has been merged. The sly cooper bug is however not a true regression and will be fixed separately in upcoming weeks. The issue can be closed if that is the only regression left.

**Comment by kd-11** [comment]
What happens if you open the RSX capture file attached? (Utilities->Open RSX capture)

**Comment by creeperjedi** [comment]
When I use current master and open the post-regression RSX capture file (I assume that's what you wanted) I get a black screen like before.
[H.A.W.X. Log - 0.0.5-7757.log](https://github.com/RPCS3/rpcs3/files/2798523/H.A.W.X.Log.-.0.0.5-7757.log)
[H.A.W.X. Log - 0.0.5-7757.log.gz](https://github.com/RPCS3/rpcs3/files/2798524/H.A.W.X.Log.-.0.0.5-7757.log.gz)
Though when I load the pre-regression RSX capture file, I get the image of the menu seen above.
(Both tests done without Write Color Buffers enabled)

**Comment by Xcedf** [comment]
Found a few regressions here
Uncharted 1 flicekring Naughty Dog intro, Simpsons The Game also may have short random artifacts during character's special attacs, which rely on WCB+CPU Blit
![5](https://user-images.githubusercontent.com/7060586/51790531-abf92e00-21a7-11e9-8fb5-a26d75ca20cd.jpg)
both regressed after 52ac0a90

**Comment by questore** [comment]
> acter's special attacs, which rely on WCB+CPU Blit
Uncharted flickering is gone. Just you use only WCB (no CPU blit)

**Comment by Xcedf** [comment]
Ok i agree, for Uncharted with WCB flickering are gone, but for Simpsons The Game CPU Blit is vital, there will be even more severe artefacts with just WCB, and and the flickering in Simpsons happens even with WCB+CPU Blit are ON, so still an issue

**Comment by legend800** [comment]
Time Crisis 4 is still broken on VK (washed out):
![2019-01-27 09_52_38-window](https://user-images.githubusercontent.com/20983844/51804831-442ffb00-221a-11e9-8faa-7ed8770405b7.png)

I can just reopen this issue for tracking:
https://github.com/RPCS3/rpcs3/issues/4949

And on OpenGL, it's completely broken, crash on load (tried clearing cache):
[RPCS3.zip](https://github.com/RPCS3/rpcs3/files/2800325/RPCS3.zip)

**Comment by MSuih** [comment]
Sorry for the late report, hadn't gotten around to test this before. In **Lollipop Chainsaw**, there's a bug where certain effects just render a static image on screen. After FBO_Fixes, this static image will cause VRAM usage to climb up rapidly every time it's shown, eventually causing an out of video memory error after just one or two of these screens.

This does not affect OpenGL, only Vulkan. Also strict rendering mode fixes the crash, but it wasn't required before FBO fixes. Enabling debug output reduces the speed of VRAM usage growth, but that's probably because it renders less frames during those freezes. It does cause the game to print a lot of errors though, but I'm not sure how much of it is just pointless noise.

If you want an easy location in game to test this then delete your old save and start a new game. The intro to first stage has a long freeze that should fill the whole VRAM in one go.

Here's what it looks like in task manager:
![image](https://user-images.githubusercontent.com/19759300/51841085-745cc400-2316-11e9-9315-ed6b45a8eab0.png)

[Log with debug output](https://github.com/RPCS3/rpcs3/files/2803541/RPCS3.log.gz)

I tested this by compiling a version with AUDIT enabled, but it didn't make any difference.

**Comment by Xcedf** [comment]
Found another regression in StrictRendering in 1ffadbe0
before:
![27](https://user-images.githubusercontent.com/7060586/52130431-2f54cc80-264b-11e9-87ff-8a8ebbadfc88.jpg)
after:
![28](https://user-images.githubusercontent.com/7060586/52130292-d08f5300-264a-11e9-9c84-fe32ce8a5556.jpg)

placing back this code uppper in file again makes things work
if (g_cfg.video.strict_rendering_mode)
{
	if (ds) ds->memory_barrier(*m_current_command_buffer);

	for (auto &rtt : m_rtts.m_bound_render_targets)
	{
		if (auto surface = std::get<1>(rtt))
		{
			surface->memory_barrier(*m_current_command_buffer);
		}
	}
}

I'd like to notice; this revert fix only affects the Strict Render, the Normal render still remain buggy, i wander if there's an equal for normal rendering.

**Comment by kd-11** [comment]
Disable the vulkan memory allocator. On AMD, there are very few memory handles available to the application, if we're leaking a temp resource, you should run out of handles long before you run out of memory. The driver only allows 4096 objects to be allocated. The allocator does huge pool allocations and makes memory allocations just in software like a normal heap and bypasses this, making it harder to determine if something is leaking.

**Comment by kd-11** [comment]
For the other graphics glitches, there are bugs that were uncovered by doing things correctly. I urge testing with this experimental build https://ci.appveyor.com/project/kd-11/rpcs3/builds/22079284, resolves some known issues reported here.

**Comment by legend800** [comment]
Tested the new build:
Time Crisis bloom regression above not fixed.
Backbreaker Vengeance (needs WCB+Blit to avoid logged regressions): Lost ~7-8 fps (30 fps title).
Ran through a bunch of other games - no improvements/regressions of note.

**Comment by creeperjedi** [comment]
Using the newer build: The H.A.W.X regression was not fixed for me

**Comment by kd-11** [comment]
Provide renderdoc before and after since RSX captures are tainted by broken graphics code. Use the latest version of renderdoc.

**Comment by MSuih** [comment]
Hmm, interesting. I experimented with this and your suggestion about disabling vulkan memory allocator didn't help, the game still ran out of VRAM just as fast as it used to. However, since the game spams "Blit request denied because the source region does not fit!" to console I tried forcing CPU blit, and that somehow fixes this issue. By enabling WCB the whole effect is rendered properly, which I wasn't aware of.

So you do have a memory leak in some part of emulator code, and it's a part that gets skipped if CPU blit is enabled.

**Comment by kd-11** [comment]
Thats a huge part with too much going on, although the source not fitting is an interesting glitch for sure. I'm curious since the hw blit path does not actually allocate anything with the exception of blit resources which are supposed to be reused whenever possible, so it is likely the old known blit dst leak. I sill wonder why it is not reusing the created resources if this is the case.

**Comment by MSuih** [comment]
It only activates when using Vulkan and Strict rendering somehow fixes it. Also the issue never happened before FBO_Fixes was merged. Maybe that would help narrowing it down?

But either way, I think I've pestered you enough about it. That issue only seems to affect one game and even then it can be avoided with settings changes, so it's not something that you should spend too much time looking into.

**Comment by kd-11** [comment]
Some of the other glitches are likely fixed by https://github.com/RPCS3/rpcs3/pull/5635 if they affect vulkan only.

**Comment by legend800** [comment]
- TC4 is still washed out and OpenGL causes fatal errors. Both fine prior to FBO fixes. I can reopen the specific TC4 issue to track separately if you want. #4949
- Performance regression still present in Backbreaker, maybe others (though not as bad as the build from a few days ago) on these new test builds/PR
- No other issues/improvements of note. 

![2019-02-05 08_39_16-window](https://user-images.githubusercontent.com/20983844/52289942-e12e1a80-2923-11e9-8c2d-e20924d6e7f2.png)

**Comment by kd-11** [comment]
There have been no new builds related to fbos, so if you still have a performance regression, look into your system setup. The new PR build does not impact performance.
As for tc4, do not reopen a duplicate. More open tickets result in slower progress since emulator development still has to move forward and minor issues have to be temporarily postponed often. Attach renderdoc dumps here instead (vulkan trace is fine).

**Comment by creeperjedi** [comment]
I tried to get an RSX Capture, but the emulator just instantly crashes when I click the "RSX Capture" button in the debugger. Instead, I got a RenderDoc, if that helps.

Tested using default settings + WRC, CPU Blit, and Use Legacy OpenGL Buffers
<details><summary>RenderDoc</summary>

<br>

<details><summary>Download RenderDoc from GitHub (multiple files):</summary>
(I had to split it up because GitHub only supports 10MB files)

When downloaded, remove the .zip off the end of the file
[H.A.W.X. RenderDoc (OpenGL) - 0.0.5-7688 (2).7z.001.zip](https://github.com/RPCS3/rpcs3/files/2834131/H.A.W.X.RenderDoc.OpenGL.-.0.0.5-7688.2.7z.001.zip)
[H.A.W.X. RenderDoc (OpenGL) - 0.0.5-7688 (2).7z.002.zip](https://github.com/RPCS3/rpcs3/files/2834130/H.A.W.X.RenderDoc.OpenGL.-.0.0.5-7688.2.7z.002.zip)
</details>
<br>

Download RenderDoc from Google Drive: 
https://drive.google.com/open?id=1YrMIlHhdKfAdwbPcn8KDgchAu9GzkP7q
</details>
<br>
Logs:

[H.A.W.X. Log - 0.0.5-7688 (2).log](https://github.com/RPCS3/rpcs3/files/2834128/H.A.W.X.Log.-.0.0.5-7688.2.log)
[H.A.W.X. Log.gz - 0.0.5-7688.log (2).gz](https://github.com/RPCS3/rpcs3/files/2834129/H.A.W.X.Log.gz.-.0.0.5-7688.log.2.gz)

**Comment by legend800** [comment]
RenderDoc for TC4 issue:
https://drive.google.com/file/d/1LT4vHX84p-4unYnCm33k-cpJ8U5T-QOy/view?usp=sharing

**Comment by kd-11** [comment]
Retest bugs with https://github.com/RPCS3/rpcs3/pull/5675

**Comment by legend800** [comment]
Time Crisis 4 still has broken bloom, like above.

**Comment by MSuih** [comment]
HW blit memory leak is still there. I also found two other games which are affected by it; NHL 14 and Uncharted 3.

**Comment by creeperjedi** [comment]
H.A.W.X now displays the main menu, just like before the regression, but freezes on this screen (see image below) right before it would enter ingame.
<img width="641" alt="h a w x freeze screen" src="https://user-images.githubusercontent.com/38361979/53700877-eec0ae00-3dc4-11e9-9b01-80ac13a3c99c.png">
[H.A.W.X. Log - 0.0.6-426355a1.log](https://github.com/RPCS3/rpcs3/files/2923609/H.A.W.X.Log.-.0.0.6-426355a1.log)
Settings: Default + Vulkan + Write Color Buffers

**Comment by creeperjedi** [comment]
@legend800 It might be worth testing Time Crisis 4 again because a recent update to the aforementioned PR fixed some broken bloom in H.A.W.X. (which I can see now, since the game doesn't just display black anymore).

<details>
<summary>Before:</summary>
<img width="641" alt="h a w x high bloom" src="https://user-images.githubusercontent.com/38361979/53701123-ee75e200-3dc7-11e9-93b9-9c32d7fbadaf.png">

</details>

<details>
<summary>After:</summary>
<img width="641" alt="h a w x normal bloom" src="https://user-images.githubusercontent.com/38361979/53701129-f766b380-3dc7-11e9-81cd-da948a273368.png">
</details>

**Comment by legend800** [comment]
Good call. Yep, all fixed now on latest commit in the new PR.

![2019-03-03 13_11_22-window](https://user-images.githubusercontent.com/20983844/53701981-26c0f480-3db7-11e9-9511-91691238e48e.png)

**Comment by kd-11** [comment]
Closing since most graphics bugs here if not all are fixed in latest master. Open new bug tickets if needed.
